### ISSUE TITLE
Error when performing helm rollback to a release containing a resource with the `helm.sh/resource-policy: keep` annotation

### ISSUE THREAD

**Issue by dudicoco** [helm/helm#8228]
Hi,
We are creating configmaps with a hash in the name, generated from the contents of the configmap data. This is done as a workaround to the k8s issue with rolling updates and configmaps.
In addition, we add the `helm.sh/resource-policy: keep` annotation to make sure that the previous configmap is not deleted during the rolling update, so we end up with several "versioned" configmaps such as:
`config-common-223ee6a5d95cc68fd4af39af3b059b595f54222acf4d47ab0b53134ce89c2cfe`

When performing `helm rollback` to a previous revision which contains a configmap name hash different than the current release revision we get the following error:
`Error: no ConfigMap with the name "config-common-223ee6a5d95cc68fd4af39af3b059b595f54222acf4d47ab0b53134ce89c2cfe" found`

Performing another helm rollback resolves this issue.

I assume the addition of https://github.com/helm/helm/pull/7649 allowed adopting the configmap back into the release, however this should happen on the first rollback attempt.

Output of `helm version`:
```
version.BuildInfo{Version:"v3.2.1", GitCommit:"fe51cd1e31e6a202cba7dead9552a6d418ded79a", GitTreeState:"clean", GoVersion:"go1.13.10"}
```
Output of `kubectl version`:
```
\Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.6", GitCommit:"96fac5cd13a5dc064f7d9f4f23030a6aeface6cc", GitTreeState:"clean", BuildDate:"2019-08-19T11:13:49Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.9-eks-502bfb", GitCommit:"502bfb383169b124d87848f89e17a04b9fc1f6f0", GitTreeState:"clean", BuildDate:"2020-02-07T01:31:02Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```
Cloud Provider/Platform (AKS, GKE, Minikube etc.): AWS

**Comment by bacongobbler** [comment]
Hi @dudicoco. I'm not sure what you are describing here. Can you please provide a set of instructions to describe your issue in more detail?

**Comment by dudicoco** [comment]
@bacongobbler, sure, it's very easy to recreate, here are the steps:

1. Create a chart with configmap named `config-1` and `helm.sh/resource-policy: keep` annotation
2. Install a release
3. Update the chart's configmap name to `config-2`
4. Upgrade the release
5. You should now have two configmaps in the cluster - `config-1` and `config-2`
6. Perform `helm rollback` to the previous revision
7. An error should occur: `Error: no ConfigMap with the name "config-1" found`
8. Perform `helm rollback` once more should finish in success

**Comment by liuming-dev** [comment]
> @bacongobbler, sure, it's very easy to recreate, here are the steps:
> 
> 1. Create a chart with configmap named `config-1` and `helm.sh/resource-policy: keep` annotation
> 2. Install a release
> 3. Update the chart's configmap name to `config-2`
> 4. Upgrade the release
> 5. You should now have two configmaps in the cluster - `config-1` and `config-2`
> 6. Perform `helm rollback` to the previous revision
> 7. An error should occur: `Error: no ConfigMap with the name "config-1" found`
> 8. Performing `helm rollback` once more should finish in success

I can reproduce this case, this case hits the logic 

https://github.com/helm/helm/blob/662090a37e977922828779d249f9cc4f8fc7ed1d/pkg/kube/client.go#L199-L203

**Comment by etedpet** [comment]
Yesterday I wrote reported #8741 which was considered a duplicate of this issue and therefor closed by @bacongobbler.

However that issue is NOT resolved by performing another helm rollback (as described in this issue).

Also it is not really related to the `helm.sh/resource-policy: keep` annotation, it was just the simplest way to reproduce the issue.

The issue is that the running K8S cluster _contains a resource_ (StatefulSet in my case) with a **name** that is part of the _previous_ chart (manifest rolling back to) but NOT part of the _current_ chart (manifest rolling back from).

In my case this happens since an operator have "taken over" control of the StatefulSet (no longer part of the chart manifest). This is regardless of whether `helm.sh/resource-policy: keep` annotation is used or not - for me it was just a way of getting a smooth migration (not restarting the StatefulSet during upgrade).

If I do the upgrade without the "smooth"-requirement, then I could let the operator _operate_ a StatefulSet with a _different name_. The "pre-upgrade" StatefulSet will then be deleted and a new StatefulSet (with a different name) will be created by the operator. Then the rollback works as expected - but I can't have a smooth upgrade to my new chart version :-(

**Comment by etedpet** [comment]
I tried the fix #8683, but it did not solve my issue (#8741). Still got the same error during rollback.

```
> helm-latest version
version.BuildInfo{Version:"v3.3+unreleased", GitCommit:"68d0483890f0a63f784f0f014e032b3a5f40e149", GitTreeState:"clean", GoVersion:"go1.15"}

> helm-latest rollback app --debug 
rollback.go:63: [debug] preparing rollback of app
rollback.go:111: [debug] rolling back app (current: v2, target: v1)
rollback.go:70: [debug] creating rolled back release for app
rollback.go:76: [debug] performing rollback of app
client.go:174: [debug] checking 1 resources for changes
rollback.go:169: [debug] warning: Rollback "app" failed: no StatefulSet with the name "test" found
Error: no StatefulSet with the name "test" found
helm.go:94: [debug] no StatefulSet with the name "test" found
helm.sh/helm/v3/pkg/kube.(*Client).Update.func1
	/local/repos/helm/pkg/kube/client.go:216
helm.sh/helm/v3/pkg/kube.ResourceList.Visit
	/local/repos/helm/pkg/kube/resource.go:32
helm.sh/helm/v3/pkg/kube.(*Client).Update
	/local/repos/helm/pkg/kube/client.go:175
helm.sh/helm/v3/pkg/action.(*Rollback).performRollback
	/local/repos/helm/pkg/action/rollback.go:165
helm.sh/helm/v3/pkg/action.(*Rollback).Run
	/local/repos/helm/pkg/action/rollback.go:77
main.newRollbackCmd.func2
	/local/repos/helm/cmd/helm/rollback.go:69
github.com/spf13/cobra.(*Command).execute
	/home/etedpet/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842
github.com/spf13/cobra.(*Command).ExecuteC
	/home/etedpet/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950
github.com/spf13/cobra.(*Command).Execute
	/home/etedpet/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main
	/local/repos/helm/cmd/helm/helm.go:93
runtime.main
	/usr/local/go/src/runtime/proc.go:204
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1374
```

**Comment by etedpet** [comment]
Created a pull request (#9068) with a very simple example on how to reproduce/test this issue.
